### PR TITLE
upstream: Prague code merge [v1.13.15, v1.14.11]

### PR DESCRIPTION
**Pull request by buddh0**
### Description

Prague code merge [v1.13.15, v1.14.11(latest)]

### Rationale

One of the preparations for activating the Prague hard fork on bsc, status
✅ compile pass
```
make all
```
✅ fix all CIs
✅ mine blocks with node-deploy script

todo:
1. code review and fix
2. various nodes and sync test
3. QA regression testing

### Example

add an example CLI or API response...

#### Major Changes
[eth: make transaction propagation paths in the network deterministic (#29034)](https://github.com/ethereum/go-ethereum/pull/29034)
[core/state: parallelise parts of state commit (#29681)](https://github.com/ethereum/go-ethereum/pull/29681)
Load trie nodes concurrently with trie updates, speeding up block import by 5-7% ([#29519](https://github.com/ethereum/go-ethereum/pull/29519), [#29768](https://github.com/ethereum/go-ethereum/pull/29768), [#29919](https://github.com/ethereum/go-ethereum/pull/29919)).
[core/vm: reject contract creation if the storage is non-empty (](https://github.com/ethereum/go-ethereum/commit/c170cc0ab0a1f60adcde80d0af8e3050ee19da93)[#28912](https://github.com/ethereum/go-ethereum/pull/28912)[)](https://github.com/ethereum/go-ethereum/commit/c170cc0ab0a1f60adcde80d0af8e3050ee19da93)
[Add state reader abstraction (#29761)](https://github.com/ethereum/go-ethereum/pull/29761)
[Stateless witness prefetcher changes (#29519)](https://github.com/ethereum/go-ethereum/pull/29519)
_not follow changes with trie_prefetcher, the implemenation in bsc is very different_
[core: use finalized block as the chain freeze indicator (](https://github.com/ethereum/go-ethereum/commit/ca473b81cbe4a96cde4e8424c49b15ab304787bb)[#28683](https://github.com/ethereum/go-ethereum/pull/28683)[)](https://github.com/ethereum/go-ethereum/commit/ca473b81cbe4a96cde4e8424c49b15ab304787bb)
_in bsc, this feature only enabled with multi-database_

#### New EIPs
[core/vm: enable bls-precompiles for Prague (](https://github.com/ethereum/go-ethereum/commit/823719b9e1b72174cd8245ae9e6f6f7d7072a8d6)[#29552](https://github.com/ethereum/go-ethereum/pull/29552)[)](https://github.com/ethereum/go-ethereum/commit/823719b9e1b72174cd8245ae9e6f6f7d7072a8d6)
[EIP-2935: Serve historical block hashes from state](https://eips.ethereum.org/EIPS/eip-2935) ([#29465](https://github.com/ethereum/go-ethereum/pull/29465))

#### Clear Up
[eth, eth/downloader: remove references to LightChain, LightSync (#29711)](https://github.com/ethereum/go-ethereum/pull/29711)
[eth/filters: remove support for pending logs(#29574)](https://github.com/ethereum/go-ethereum/pull/29574)
[Drop large-contract (500MB+) deletion DoS protection from pathdb post Cancun (#28940)](https://github.com/ethereum/go-ethereum/pull/28940).
Remove totalDifficulty field from RPC, in accordance with [spec update](https://github.com/ethereum/execution-apis/pull/570), [#30386](https://github.com/ethereum/go-ethereum/pull/30386)

#### Merged but Reverted
[consensus, cmd, core, eth: remove support for non-merge mode of operation (](https://github.com/ethereum/go-ethereum/commit/f4d53133f6e4b13f0dbcfef3bc45e9650d863b73)[#29169](https://github.com/ethereum/go-ethereum/pull/29169)[)](https://github.com/ethereum/go-ethereum/commit/f4d53133f6e4b13f0dbcfef3bc45e9650d863b73)
[miner: refactor the miner, make the pending block on demand (](https://github.com/ethereum/go-ethereum/commit/d8e0807da22eb922539d15b0d5d01ccdd58b1267)[#28623](https://github.com/ethereum/go-ethereum/pull/28623)[)](https://github.com/ethereum/go-ethereum/commit/d8e0807da22eb922539d15b0d5d01ccdd58b1267)
[miner: lower default min miner tip from 1 gwei to 0.001 gwei(#29895)](https://github.com/ethereum/go-ethereum/pull/29895)
_bsc only has tip, 1 Gwei is the min value now_
[eth/downloader: purge pre-merge sync code (](https://github.com/ethereum/go-ethereum/commit/45baf21111c03d2954c81fdf828e630a8d7b05c1)[#29281](https://github.com/ethereum/go-ethereum/pull/29281)[)](https://github.com/ethereum/go-ethereum/commit/45baf21111c03d2954c81fdf828e630a8d7b05c1)
[all: remove forkchoicer and reorgNeeded (](https://github.com/ethereum/go-ethereum/commit/b0b67be0a2559073c1620555d2b6a73f825f7135)[#29179](https://github.com/ethereum/go-ethereum/pull/29179)[)](https://github.com/ethereum/go-ethereum/commit/b0b67be0a2559073c1620555d2b6a73f825f7135)

##### deprecated flags
- light.serve
- light.ingress
- light.egress
- light.maxpeers
- light.nopruning
- light.nosyncserve
- metrics.expensive
##### delete fields in config.toml
[Eth]
LightPeers
[Eth.GPO]
Default

#### Others
[all: update to go version 1.23.0 (#30323)](https://github.com/ethereum/go-ethereum/pull/30323)
[Switch to using Go's native log/slog package instead of golang/exp (#29302)](https://github.com/ethereum/go-ethereum/pull/29302).
[Add the geth db inspect-history command to inspect pathdb state history (#29267)](https://github.com/ethereum/go-ethereum/pull/29267).
Improve the discovery protocol's node revalidation ([#29572](https://github.com/ethereum/go-ethereum/pull/29572), [#29864](https://github.com/ethereum/go-ethereum/pull/29864), [#29836](https://github.com/ethereum/go-ethereum/pull/29836)).
[Blobpool related flags in Geth now actually work. (#30203)](https://github.com/ethereum/go-ethereum/pull/30203)
[core/rawdb: implement in-memory freezer (#29135)](https://github.com/ethereum/go-ethereum/commit/f46c878441e2e567e8815f1e252a38ad0ffafbc2)

more details to refer [release note between [v1.13.15, v1.14.11] in geth](https://github.com/ethereum/go-ethereum/releases)

### CODE REVIEW

**Comment by buddh0** [comment]
ignore Lint Commit CI, because lots of commits upstreamed

File: core/tracing/hooks.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+	// Note that system call happens outside normal transaction execution, so the `OnTxStart/OnTxEnd` hooks
+	// will not be invoked.
+	OnSystemCallStartHook = func()
+
+	// OnSystemCallEndHook is called when a system call has finished executing. Today,
+	// this hook is invoked when the EIP-4788 system call is about to be executed to set the
+	// beacon block root.
+	OnSystemCallEndHook = func()
+
+	// TODO(Nathan,Eric): refine this func
```
Comment (buddh0) [inline core/tracing/hooks.go:153]: a tmp modification
@zlacfzy please help handle `CaptureSystemTxEnd`

File: p2p/discover/table.go [hunk 2]
```diff
@@ -600,177 +507,223 @@ func (tab *Table) filterNode(n *node) bool {
...
+		wn.livenessChecks = 1
+		wn.isValidatedLive = true
 	}
-}
 
-// delete removes an entry from the node table. It is used to evacuate dead nodes.
-func (tab *Table) delete(node *node) {
-	tab.mutex.Lock()
-	defer tab.mutex.Unlock()
+	// TODO(Matus): fix the filterNode feature
```
Comment (buddh0) [inline p2p/discover/table.go:573]: p2p module changes a lot,
I tried to put filterNode logic into new codes, but failed
@MatusKysel please help recover the `filterNode` feature

File: triedb/pathdb/difflayer.go [hunk 3]
```diff
@@ -232,57 +231,15 @@ func (dl *diffLayer) node(owner common.Hash, path []byte, hash common.Hash, dept
...
-				return nil, newUnexpectedNodeError("diff", hash, n.Hash, owner, path, n.Blob)
-			}
 			dirtyHitMeter.Mark(1)
 			dirtyNodeHitDepthHist.Update(int64(depth))
 			dirtyReadMeter.Mark(int64(len(n.Blob)))
-			return n.Blob, nil
+			return n.Blob, n.Hash, &nodeLoc{loc: locDiffLayer, depth: depth}, nil
 		}
 	}
+	//TODO(will-2012): https://github.com/bnb-chain/bsc/pull/2508 broken
```
Comment (buddh0) [inline triedb/pathdb/difflayer.go:240]: @will-2012 metioned that
the performance by PR https://github.com/bnb-chain/bsc/pull/2508 has been roll back,
for intefaces changed

File: cmd/geth/testdata/config.toml [hunk 4]
```diff
@@ -3,7 +3,6 @@ NetworkId = 714
 SyncMode = "full"
 NoPruning = false
 NoPrefetch = false
-LightPeers = 100
```
Comment (buddh0) [inline cmd/geth/testdata/config.toml:6]: @zzzckck 
please remove this field in https://github.com/bnb-chain/bsc/releases/download/v1.4.15/mainnet.zip
and https://github.com/bnb-chain/bsc/releases/download/v1.4.15/testnet.zip

File: triedb/pathdb/difflayer.go (hunk 3 above)
Comment (will-2012) [reply triedb/pathdb/difflayer.go]: PBSS hashcache has been re-enable by PR https://github.com/bnb-chain/bsc/pull/2765, pls review~

**Comment by maoueh** [comment]
Re-linking my comment https://github.com/bnb-chain/bsc/pull/2768#issuecomment-2498819428 here for increased visibility.

**Review by zlacfzy — APPROVED** [review]

**Review by galaio — APPROVED** [review]

**Review by zzzckck — APPROVED** [review]

**Pull request merged**